### DISCUSSION TITLE
【教程贴】企业版Docker镜像构建方法交流讨论

### DISCUSSION THREAD

**Discussion by WenkaiZhou** [lsky-org/lsky-pro#749] [教程]
搭建企业版时，发现没有Docker镜像部署的相关资料，浅浅研究了下，面向小白，大佬轻拍～

1. 跳过在服务器安装Docker环境；

2. `clone` [Docker-LskyProEnterprise](https://github.com/WenkaiZhou/Docker-LskyProEnterprise) 仓库；

3. 官方下载企业版压缩包，重命名为lsky-pro.zip放在代码目录下；

4. 构建镜像，版本可以随意，建议和下载的LskyPro版本一致；

        docker build -t lsky-pro:1.5.1 .

5. 为了方便不同服务共享数据库服务，`MySql`也使用Docker容器的方式，首先创建网络；

        docker network create -d bridge --subnet=172.18.0.0/16 --gateway=172.18.0.1 web_network

6. 创建`MySql`容器，记得把两个密码参数换下；

        docker run -d \
            --name=mysql \
            --restart unless-stopped \
            --network=web_network \
            --ip=172.18.0.2 \
            -p 3306:3306 \
            -v /docker/mysql:/var/lib/mysql \
            -e MYSQL_ROOT_PASSWORD=ROOT密码 \
            -e MYSQL_USER=lsky-pro \
            -e MYSQL_PASSWORD=lsky-pro密码 \
            -e MYSQL_DATABASE=lsky-pro \
            mysql:5.7.9

7.  创建`lsky-pro`容器，记得把你的授权替换下：

        docker run -d \
            --name=lsky-pro \
            --restart unless-stopped \
            --network=web_network \
            --ip=172.18.0.3 \
            -p 80:80 \
            -e APP_SERIAL_NO=你的许可证编号 \
            -e APP_SECRET=你的许可证密钥 \
            -e APP_URL=你的站点域名 \
            -v /docker/lsky-pro:/var/www/html \
            lsky-pro:1.5.1

8. 访问站点，在管理后台进行设置。

如果觉得要配置网络和数据库容器麻烦，也可以直接使用`docker-compose`的方式；

当然，此时服务是`http`的，也可以再开一个`nginx`容器进行反向代理，为外部提供`https`服务。

**Comment by defpis** [comment]
请问一下自己构建的镜像，请求install 接口500报错，错误日志: 
Failed to open stream: Permission denied in /var/www/html/vendor/laravel/framework/src/Illuminate/Filesystem/Filesystem.php:190
然后chmod 777 vendor/ 也还是不行

**Reply by WenkaiZhou** [reply]
我之前使用的apache作为webserver后来切换为nginx，没有再走安装的步骤。试了下，确实有你说的这个问题。是由于没有修改nginx的用户为www导致的，修改已经提交了，麻烦你再试下，有问题的话再沟通。

**Reply by defpis** [reply]
我试了应该还是有问题，运行时创建的.env文件是root:root会导致写入问题，我进docker 整体修改一下权限就好了，但是这样打出来的docker还是没法直接用

**Reply by defpis** [reply]
最好让php动态生成的文件跟随chown www:www 和cmod 755的权限，不然很容易出问题

**Reply by defpis** [reply]
还有一个头疼的问题就是 check heath 的脚本可能有问题，导致虽然正常运行但是识别到错误的状态：
<img width="615" alt="image" src="https://github.com/lsky-org/lsky-pro/assets/48018016/588595d0-82ac-4312-8f0f-b061a9dbf5bf">

**Reply by defpis** [reply]
或者你直接分享一下apache的docker构建流程，可能apache和php兼容性更好？没写过php不太懂，我在虚拟机Ubuntu用Nginx完整部署，权限问题也会有，出了问题chmod和 chown能解决，但是还是希望有一个docker能直接用，不需要进入docker内部修改内容。

**Reply by WenkaiZhou** [reply]
针对你说的问题，我近期再看下，apache的镜像会大很多，目前这个方案的镜像比较小，而且执行效率优于apache。根据你的现象还存在下面两个问题？
1. .env 的用户非 www:www，权限非 755；
2. check heath 运行有问题；

**Comment by defpis** [comment]
没有测试的，各种问题，还是不建议用这个配置

**Reply by WenkaiZhou** [reply]
很抱歉给你带来不好的体验，创建本帖的初衷就是相互讨论交流，大家共同维护可用的docker版本。
又对权限部分进行了收敛，统一用 `www-data` 用户来启动 `supervisord`，希望以此来解决运行异常和以`root`运行的不安全问题。

**Comment by defpis** [comment]
后面我尝试过修改的版本，上传图片会有403问题，我没有继续尝试，不确定最新版本是否还有问题。这里分享一下我尝试过没问题的部署步骤：
### 构建企业版镜像
1. 确保机器上安装了docker
	```shell
	docker -h
	```
2. `git clone https://github.com/WenkaiZhou/Docker-LskyProEnterprise` 仓库到本地，`git checkout 37b1f406f52ecc8ea45f9777d233510f0fbf8ae1` 切换到 `Update README` 这个commit，修改一下文件：
	- `Dockerfile`：注释`HEALTHCHECK`，因为会导致健康检查异常，打很多日志
	```Dockerfile
	- HEALTHCHECK --timeout=10s CMD curl --silent --fail http://127.0.0.1:9000/fpm-ping
	+ # HEALTHCHECK --timeout=10s CMD curl --silent --fail http://127.0.0.1:9000/fpm-ping
	```
	- `entrypoint.sh`：修改`755`为`777`，因为可能会有权限问题
	```
	- chmod -R 755 /var/www/html/
	+ chmod -R 777 /var/www/html/
	```
1. 登录 https://www.lsky.pro/login 下载源码放到 `Docker-LskyProEnterprise` 目录下，修改文件名为 `lsky-pro.zip`
2. 执行 `docker build -t lsky-pro:1.5.1 .` 构建镜像
3. `docker save -o lsky-pro.tar lsky-pro:1.5.1` 保存镜像到本地，上传 `lsky-pro.tar`到其他机器上执行`docker load -i lsky-pro.tar`载入镜像
### 极空间部署
#### 部署MariaDB数据库
1. 拉取 `mariadb:10.8.2` 镜像，创建容器
	1. 设置目录 `/xx/mariadb`  挂载 `/var/lib/mysql`
	2. 设置端口 `33306` 映射 `3306`
	3. 设置环境变量 `MARIADB_ROOT_PASSWORD` 为 `root` 
2. 进入容器内(docker exec -it <container_name> bash)，`mysql -uroot -proot`连接数据库，创建数据库和用户并授权
	```SQL
	# 查看已有用户和主机
	SELECT User,Host FROM mysql.user;
	# 创建用户和主机，用%表示允许任何主机连接
	CREATE USER 'lsky_pro'@'%' IDENTIFIED BY 'lsky_pro';
	# 创建数据库
	CREATE DATABASE lsky_pro;
	# 授予mydb数据库访问权限，用*表示所有库或所有表
	GRANT ALL PRIVILEGES ON lsky_pro.* TO 'lsky_pro'@'%';
	# 更新权限信息，使之生效
	FLUSH PRIVILEGES;
	```
#### 部署LskyPro图床
1. 上传镜像并导入，创建容器
	1. 设置目录 `/xx/lsk-pro`  挂载 `/var/www/html`
	2. 设置端口 `8800` 映射 `80`
	3. 添加环境变量，登录 https://www.lsky.pro/login 获取
		1. APP_SERIAL_NO
		2. APP_SECRET
		3. APP_URL
2. 启动镜像，访问`[机器IP:8800]`，会等待一段时间初始化
3. 出现安装引导页面后先不要继续，进入容器内部执行命令：
```
chmod -R 777 /var/www/html
chown -R www-data:www-data /var/www/html
```
4. 数据库URL填写 `[机器IP:33306]`，数据库密码账号都是`lsky_pro`
5. 继续安装直到完成

**Reply by WenkaiZhou** [reply]
我使用的新版本目前还没有发现问题，有兴趣的话可以尝试下(要先做好数据的备份，毕竟数据无价)

**Reply by Xm798** [reply]
为了解决权限问题和 crontab 定时任务问题，基于 lsio 的 baseimage 制作了一个镜像 https://github.com/Xm798/docker-lsky-pro-enterprise 可以测试

**Comment by heiyuan0801** [comment]
可以发到 lskypro.com
